File: spark/src/main/scala/geotrellis/spark/io/file/FileLayerReader.scala
/*
 * Copyright 2016 Azavea
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package geotrellis.spark.io.file

import geotrellis.tiling._
import geotrellis.layers.LayerId
import geotrellis.layers._
import geotrellis.layers.file.{FileAttributeStore, FileLayerHeader, KeyPathGenerator}
import geotrellis.layers.avro._
import geotrellis.layers.index._
import geotrellis.spark._
import geotrellis.spark.io.FilteringLayerReader
import geotrellis.util._

import com.typesafe.scalalogging.LazyLogging

import org.apache.spark.SparkContext

import spray.json.JsonFormat

import scala.reflect.ClassTag

/**
 * Handles reading raster RDDs and their metadata from a filesystem.
 *
 * @param attributeStore  AttributeStore that contains metadata for corresponding LayerId
 * @tparam K              Type of RDD Key (ex: SpatialKey)
 * @tparam V              Type of RDD Value (ex: Tile or MultibandTile )
 * @tparam M              Type of Metadata associated with the RDD[(K,V)]
 */
class FileLayerReader(
  val attributeStore: AttributeStore,
  catalogPath: String
)(implicit sc: SparkContext) extends FilteringLayerReader[LayerId] with LazyLogging {

  val defaultNumPartitions = sc.defaultParallelism

  def read[
    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
    V: AvroRecordCodec: ClassTag,
    M: JsonFormat: Component[?, Bounds[K]]
  ](id: LayerId, tileQuery: LayerQuery[K, M], numPartitions: Int, filterIndexOnly: Boolean) = {
    if(!attributeStore.layerExists(id)) throw new LayerNotFoundError(id)

    val LayerAttributes(header, metadata, keyIndex, writerSchema) = try {
      attributeStore.readLayerAttributes[FileLayerHeader, M, K](id)
    } catch {
      case e: AttributeNotFoundError => throw new LayerReadError(id).initCause(e)
    }

    val layerPath = header.path

    val queryKeyBounds = tileQuery(metadata)
    val layerMetadata = metadata.setComponent[Bounds[K]](queryKeyBounds.foldLeft(EmptyBounds: Bounds[K])(_ combine _))
    val maxWidth = Index.digits(keyIndex.toIndex(keyIndex.keyBounds.maxKey))
    val keyPath = KeyPathGenerator(catalogPath, layerPath, maxWidth)
    val decompose = (bounds: KeyBounds[K]) => keyIndex.indexRanges(bounds)
    val rdd = FileRDDReader.read[K, V](keyPath, queryKeyBounds, decompose, filterIndexOnly, Some(writerSchema), Some(numPartitions))

    new ContextRDD(rdd, layerMetadata)
  }
}

object FileLayerReader {
  def apply(attributeStore: AttributeStore, catalogPath: String)(implicit sc: SparkContext): FileLayerReader =
    new FileLayerReader(attributeStore, catalogPath)

  def apply(catalogPath: String)(implicit sc: SparkContext): FileLayerReader =
    apply(new FileAttributeStore(catalogPath), catalogPath)

  def apply(attributeStore: FileAttributeStore)(implicit sc: SparkContext): FileLayerReader =
    apply(attributeStore, attributeStore.catalogPath)
}

File: geotrellis-python/geotrellis/spark/io/file/FileLayerReader.py
from __future__ import absolute_import
from geotrellis.spark.io.FilteringLayerReader import FilteringLayerReader
from geotrellis.spark.io.package_scala import LayerNotFoundError, AttributeNotFoundError
from geotrellis.spark.io.index.Index import Index
from geotrellis.spark.io.file.FileLayerHeader import FileLayerHeader
from geotrellis.spark.io.file.KeyPathGenerator import generate_key_path_func
from geotrellis.spark.io.file.FileRDDReader import FileRDDReader
from geotrellis.spark.io.file.FileAttributeStore import FileAttributeStore
from geotrellis.spark.ContextRDD import ContextRDD
from geotrellis.spark.KeyBounds import KeyBounds

def _get_params(first, second):
    if second is None:
        if isinstance(first, FileAttributeStore):
            return first, first.catalogPath
        elif isinstance(first, str):
            return FileAttributeStore(first), first
        else:
            raise Exception("wrong params ({0}, {1})".format(str(first), str(second)))
    else:
        return first, second

class FileLayerReader(FilteringLayerReader):
    def __init__(self, sc, attributeStore, catalogPath = None):
        attribute_store, catalog_path = _get_params(attributeStore, catalogPath)
        self.attributeStore = attribute_store
        self._catalogPath = catalog_path
        self._sc = sc

    @property
    def defaultNumPartitions(self):
        return self._sc.defaultParallelism

    def _read3(self, K, V, M, _id, rasterQuery, numPartitions, filterIndexOnly):
        if not self.attributeStore.layerExists(_id):
            raise LayerNotFoundError(_id)
        attrs = None
        try:
            attrs = self.attributeStore.readLayerAttributes(FileLayerHeader, M, K, _id)
        except AttributeNotFoundError as e:
            raise LayerReadError(_id, cause = e)
        header = attrs.header
        metadata = attrs.metadata
        keyIndex = attrs.keyIndex
        writerSchema = attrs.schema

        layerPath = header.path
        queryKeyBounds = rasterQuery(metadata)
        maxWidth = Index.digits(keyIndex.toIndex(keyIndex.keyBounds.maxKey))
        keyPath = generate_key_path_func(self._catalogPath, layerPath, maxWidth)
        decompose = lambda bounds: keyIndex.indexRanges(KeyBounds.toTuple(bounds))
        rdd = FileRDDReader(self._sc).read(K, V, keyPath, queryKeyBounds, decompose, filterIndexOnly, writerSchema, numPartitions)
        return ContextRDD(rdd, metadata)
